File: Chapter 02/ProxyDesignPatternExample.java
interface Picture {
    void display();
}


class RealPicture implements Picture {
    private String filename;

    public RealPicture(String filename) {
        this.filename = filename;
        loadFromDisk();
    }
    private void loadFromDisk() {
        System.out.println("Loading picture: " + filename);
    }
    public void display() {
        System.out.println("Displaying picture: " + filename);
    }
}


class ProxyPicture implements Picture {
    private RealPicture realPicture;
    private String filename;

    public ProxyPicture(String filename) {
        this.filename = filename;
    }

    public void display() {
        if (realPicture == null) {
            realPicture = new RealPicture(filename);
        }
        realPicture.display();
    }
}


public class ProxyDesignPatternExample {
    public static void main(String[] args) {
        Picture picture1 = new ProxyPicture("picture1.png");
        Picture picture2 = new ProxyPicture("picture2.png");
        picture1.display();
        picture1.display();
        picture2.display();
    }
}

File: Chapter 02/ProxyDesignPatternExample.py
class Picture:
    def display(self):
        pass
class RealPicture(Picture):
    def __init__(self, filename):
        self.filename = filename
        self.load_from_disk()
    def load_from_disk(self):
        print(f"Loading picture: {self.filename}")
    def display(self):
        print(f"Displaying picture: {self.filename}")
class ProxyPicture(Picture):
    def __init__(self, filename):
        self.filename = filename
        self.real_picture = None
    def display(self):
        if self.real_picture is None:
            self.real_picture = RealImage(self.filename)
        self.real_picture.display()
if __name__ == "__main__":
    picture1 = ProxyPicture("picture1.png")
    picture2 = ProxyPicture("picture2.png")
    picture1.display()
    picture1.display()
    picture2.display()
